capital/code.ts

```diff
@@ -0,0 +1,51 @@
+// This plugin fetches a list of countries from an external API and shows the capital of the selected country
+
+let countries, resultObj;
+
+// Create an invisible iframe UI to use network API's
+figma.showUI(
+  `<script>
+    (async (event) => {
+      const res = await fetch("https://api.sampleapis.com/countries/countries");
+      const json = await res.json();
+      window.parent.postMessage({ pluginMessage: json }, "*");
+    })();
+  </script>`,
+  { visible: false }
+);
+
+figma.ui.onmessage = (json) => {
+  countries = json.map((country, index) => {
+    return {
+      name: country.name,
+      data: {
+        index: index,
+        name: country.name,
+        capital: country.capital,
+      },
+    };
+  });
+  // Set the suggestions after the a message has been received from the iframe
+  resultObj.setSuggestions(countries);
```

Comment:
I think there are a couple possible race conditions here:
- If you type a character before the countries are loaded, the countries will never be loaded for that input event
- If the countries are loaded before the first input event, then they will never be loaded for that input event.

I think the safer way of doing this is to make `countries` a promise that can be awaited in the `input` event handler. Something like:

```ts
let resolveCountries = (countries: string[]) => {}
const countriesPromise = new Promise<string[]>((resolve) => {resolveCountries = resolve})
// ...
figma.ui.onmessage = (json) => {
  // ...
 resolveCountries(countries)
}
// on('input'):
result.setLoadingMessage('Loading countries...');
const countries = await countriesPromise
result.setSuggestions(/* filter countries */)
```

Reply:
Thanks! I'll update as you've suggested.